### PR TITLE
Dashboard multisite migrate

### PR DESCRIPTION
**Pull request by aaSharma14**
Screenshot from 2023-04-25 15-56-18
Screenshot from 2023-04-25 16-13-44
Screenshot from 2023-04-25 16-09-48

<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by cloudbehl** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts [hunk 1]
```diff
@@ -42,13 +43,15 @@ export class RgwMultisiteDetailsComponent implements OnDestroy, OnInit {
 
   messages = {
     noDefaultRealm: $localize`Please create a default realm first to enable this feature`,
-    noMasterZone: $localize`Please create a master zone for each zonegroups to enable this feature`
+    noMasterZone: $localize`Please create a master zone for each zonegroups to enable this feature`,
+    disableMigrate: $localize`Migration to multi-site already done`
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts:47]:
```suggestion
    disableMigrate: $localize`Deployment is already migrated to multi-site system.`
```

**Review by nizamial09 — APPROVED** [review]
LGTM! minor nit-

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts (hunk 1 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts]: Done. Thanks @nizamial09

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.html [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                *ngIf="multisiteMigrateForm.showError('zonegroup_endpoints', formDir, 'required')"
+                i18n>This field is required.</span>
+          <span class="invalid-feedback"
+                *ngIf="multisiteMigrateForm.showError('zonegroup_endpoints', formDir, 'endpoint')"
+                i18n>Please enter a valid IP address.</span>
+        </div>
+      </div>
+      <div class="form-group row">
+        <label class="cd-col-form-label required"
+               for="zonegroupName"
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.html]: Done

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts [hunk 3]
```diff
@@ -312,6 +336,23 @@ export class RgwMultisiteDetailsComponent implements OnDestroy, OnInit {
     }
   }
 
+  getDisableMigrate() {
+    if (
+      this.realms.length === 0 &&
+      this.zonegroups.length === 1 &&
+      this.zonegroups[0].name === 'default' &&
+      this.zones.length === 1 &&
+      this.zones[0].name === 'default'
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts:345]: The button to migrate gets disabled if the name of the zonegroup and zone is not `default`? 
Why is that?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+          })
+        ]
+      }),
+      zone_endpoints: new FormControl([], {
+        validators: [
+          CdValidators.custom('endpoint', (value: string) => {
+            if (_.isEmpty(value)) {
+              return false;
+            } else {
+              if (value.includes(',')) {
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-migrate/rgw-multisite-migrate.component.ts:90]: Zone endpoints is marqued as a required field, but it doesn't seem to be required

image

**Review by Pegonzal — COMMENTED** [review]
I don't understand what are we accomplishing with this process. As far as I understood, the only thing we are doing here is creating a realm, adding the previous zonegroup and zone to it and renaming it.

Following the documentation to complete the migration from a 
single-site to a multi-site (https://docs.ceph.com/en/quincy/radosgw/multisite/#migrating-a-single-site-deployment-to-multi-site) there are two important steps left:

Restart the Ceph Object Gateway:
   `systemctl restart ceph-radosgw@rgw.'hostname -s'`
   
  And
  
  proceed to [Configure a Secondary Zone](https://docs.ceph.com/en/quincy/radosgw/multisite/#configure-secondary-zones) and create a secondary zone in the master zonegroup.
  
As it is currently, we are still left with a single zonegroup and zone in our "multi-site" configuration. So there is nothing of multi here.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts (hunk 3 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts:345]: You need a realm to enable the Migrate button, screenshot shows you do not have one.

Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts:345]: Why do we need a realm to enable a migrate button. If we have Zone group and a Zone/s at that time we need a migrate button and once we have a Realm -> Zonegroup and Zones available available we can assume we have migrated.

Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts:345]: Ohh! I mis-understood this to Import/export.! @cloudbehl  Yes you are correct. To answer @Pegonzal question - Yes, as per our implementation, we are allowing the user to Migrate from the very initial deployment that only has `default` zonegroup and zone.

**Comment by aaSharma14** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts (hunk 3 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts:345]: Then, @aaSharma14, shouldn't we be checking if the `zone` and `zonegroup` has the `default` property? Currently, we are checking  if their name is `default`, right? I'm not undersatanding this

**Review by Pegonzal — APPROVED** [review]

**Comment by aaSharma14** [comment]
jenkins test make check

**Pull request merged**